### ISSUE TITLE
Question about the balance loss of MoE

### ISSUE THREAD

**Issue by guangqianyang** [aaronhan223/Multimodal-Transformer#6]
Thanks for your wonderful work. I have one question: You mentioned that FuseMoE used MoE and the corresponding balance loss for balancing the selection of experts in MoE. However, in your code, I found that it did not return the balance loss calculated by the MoE module, as follows:
In 623-653 in core/module.py.
``` 
        if self.args.cross_method in ["moe", "hme"]:
            x_mod_in = [torch.reshape(x, (bs, -1)) for x in x_list]
            embd_len_list = [0] + list(np.cumsum([x.shape[1] for x in x_mod_in]))
            embeddings = torch.concat(x_mod_in, dim=1)
            if torch.isnan(embeddings).any():
                return None
            moe_out, balance_loss = self.moe(x_mod_in, modalities=modality)
            x_mod_out = [moe_out[:, embd_len_list[i]:embd_len_list[i + 1]] for i in range(len(embd_len_list) - 1)]
            x_allmod_output = [torch.reshape(x, (seq_len, bs, -1)) for x in x_mod_out]
            moe_output = [F.dropout(x, p=self.res_dropout, training=self.training) for x in x_allmod_output]
            x_list = [r + x for r, x in zip(residual, moe_output)]

        if self.args.cross_method == "self_cross":
            assert self.num_modalities == 2, 'Input modality should be 2 if using cross attention method.'
            x_txt, x_ts = x_list #proj_x_txt, proj_x_ts
            x_ts_to_txt, _ = self.cross_attn_1(query=x_txt, key=x_ts, value=x_ts)
            x_txt_to_ts, _ = self.cross_attn_2(query=x_ts, key=x_txt, value=x_txt)

            x_ts_to_txt = F.dropout(x_ts_to_txt, p=self.res_dropout, training=self.training)
            x_txt_to_ts = F.dropout(x_txt_to_ts, p=self.res_dropout, training=self.training)
            x_list = [r + x for r, x in zip(residual, (x_ts_to_txt, x_txt_to_ts))]

        # FNN
        residual = x_list
        x_list = [l(x) for l, x in zip(self.pre_ffn_layer_norm, x_list)]
        x_list = [F.relu(l(x)) for l, x in zip(self.fc1, x_list)]
        x_list = [F.dropout(x, p=self.relu_dropout, training=self.training) for x in x_list]
        x_list = [l(x) for l, x in zip(self.fc2, x_list)]
        x_list = [F.dropout(x, p=self.res_dropout, training=self.training) for x in x_list]
        x_list = [r + x  for r, x in zip(residual, x_list) ]
        return x_list
```
May I know where you calculate the gradient of balance loss?

**Comment by aaronhan223** [comment]
Thank you so much for your comments! We identified that this mistake was caused by a version control issue. The codebase has since been updated to correct it and address the issue you mentioned.
